google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/GetProductBiddingCategoryConstant.java

```diff
@@ -0,0 +1,193 @@
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   */
+  private static void runExample(GoogleAdsClient googleAdsClient, long customerId) {
+    // Creates the query.
+    String query =
+        "SELECT "
+            + "product_bidding_category_constant.localized_name, "
+            + "product_bidding_category_constant.product_bidding_category_constant_parent "
+            + "FROM "
+            + "product_bidding_category_constant "
+            + "WHERE "
+            + "product_bidding_category_constant.country_code IN ('US')";
+
+    // Creates the request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setPageSize(PAGE_SIZE)
+            .setCustomerId(Long.toString(customerId))
+            .setQuery(query)
+            .build();
+
+    // Creates the Google Ads Service Client.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+
+      // Creates a list of top level category nodes.
+      List<CategoryNode> rootCategories = new ArrayList<>();
+      // Creates a map of category ID to category node for all categories found in the results.
+      HashMap<String, CategoryNode> biddingCategories = new HashMap<>();
+
+      // Performs the search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      for (GoogleAdsRow googleAdsRow : response.iterateAll()) {
+        ProductBiddingCategoryConstant productBiddingCategory =
+            googleAdsRow.getProductBiddingCategoryConstant();
+
+        String name = productBiddingCategory.getLocalizedName().getValue();
+        String resourceName = productBiddingCategory.getResourceName();
+        CategoryNode node = biddingCategories.get(resourceName);
+        if (node == null) {
+          node = new CategoryNode(resourceName, name);
+          biddingCategories.put(resourceName, node);
+        } else if (node.name == null) {
+          // Ensures that the name attribute for the node is set. Name will be null for nodes added
+          // to biddingCategories as a result of being a parentNode below.
+          node.name = name;
+        }
+
+        if (productBiddingCategory.hasProductBiddingCategoryConstantParent()) {
+          String parentResourceName =
+              productBiddingCategory.getProductBiddingCategoryConstantParent().getValue();
+          CategoryNode parentNode = biddingCategories.get(parentResourceName);
+          if (parentNode == null) {
+            parentNode = new CategoryNode(parentResourceName);
+            biddingCategories.put(parentResourceName, parentNode);
+          }
+          parentNode.children.add(node);
+        } else {
+          rootCategories.add(node);
+        }
+      }
+      displayCategories(rootCategories, "");
```

Comment:
How long is the output here? Is this useful or is there a better way to present?

Reply:
The output has n lines where n is equal to # of unique top level categories + # of unique 2nd level categories + # of unique 3rd level categories + # of unique n level categories...  However, if there is only 1 unique 2nd level category, for example, a unique row wouldn't print for the top level (which would be redundant).

I've run this on an example with a large number of products/sub-products and find the organization to be quite useful and intuitive. Because of how the nodes are created, the lines of each node/sub-node are list in an intuitive order as well.